### ISSUE TITLE
Password recommendations based on custom options

### ISSUE THREAD

**Issue by flo-mic** [johannes-schliephake/nextcloud-passwords-ios#27]
Hello @johannes-schliephake
first I would like to say your ios app is great.

Probably this app and the great web extension for chrome are two main reasons for migrating from lastpass to nextcloud passwords. So great work and many thanks.

I just created a pull request for the web extension app that allows the user to define a different search behavior for password recommendations. 
https://github.com/marius-wieschollek/passwords-webextension/pull/151

I would like to see such a feature also in the ios app but have no experience with coding on ios apps. Do you think such an option inside the ios app would be possible?

The reason why I ask for this feature: I'm using an LDAP server and so serval hosts on one domain use the same credentials. Currently when I broswse to a subdomain I will not see the credentials stored for the main domain.

**Comment by johannes-schliephake** [comment]
Thanks for your positive feedback!

Your request is perfectly timed—I already wanted to merge the search and suggestion algorithms in the near future. Both are really bare-bones at the moment and, more often than not, produce almost unusable results.

However, after a few days of thinking about this issue, I came to a different solution to approach the problem. I think it it possible to achieve even better results with a new and improved algorithm, eliminating the need for seperate options (and the user having to set those options).

Essentially, I am thinking about a ranking algorithm which assigns a score to each entry based on the search term. Then, the top five (or so) entries that exceed a minimum score are selected. This should always show the "perfect match" (e.g. if the URL is exactly the same as the input), but at the same time allow for subdomains to rank high enough to be shown as well (and vice versa). I'm probably not the first person to think about such an algorithm, so i guess there should be proven implementations available.

For even more specific cases, it might be useful to add a little bit of intelligence on top: I'm thinking of additionally showing results that have been selected by the user in the past.

What do you think about this approach? Would this work for your use case? Can you think of a situation where this solution wouldn't work?

**Comment by flo-mic** [comment]
Your idea sounds great and it also solves the issue that the user would need to define his search behavior. So it is equal for everyone. Probably I can do the same on the web extension.

One additional information for you. I'm just workig on a server side setting for domain mappings. Domain mappings allow to define some urls that share the same credentials. e.g.

- apple.com / icloud.com
- microsoft.com / office.com / live.com
- ...

I will implement 3 server side settings for this. One setting contains an array of default mappings which are rolled our with each nextcloud update, this setting will be read only. Additionally to this the server admin can define own domain mapping rules and can store them in another setting as array. And the third setting will be an boolean that defines if the default domain mapping array should be considered or not. By default it will be true.

The implementation is not completed but you will have something like this in the password admin settings of your nextcloud instance.
![image](https://user-images.githubusercontent.com/52607335/109787235-4db49a00-7c0e-11eb-9653-d5d6a7c3f91e.png)

The ios app / browser extension can then use this search mapping arrays to provide even better results. What do you think is this a good idea to include as well?

**Comment by johannes-schliephake** [comment]
As long as the domain mappings are retrievable through the API I can of course implement the same logic for the iOS app.

I think it is also possible to specify multiple URLs for an entry by adding custom fields with additional URLs in the web app. It would be a good opportunity to add support for this behavior in the iOS app as well.

I will probably implement our ideas one at a time. I think the search/suggestion algorithms are the most important for now.

**Comment by johannes-schliephake** [comment]
I have now implemented a much improved search and suggestions algorithm. Apart from just being a better experience, it also includes what we have discussed here, i.e. matching domains with subdomains and vice versa.

Although I have closed this issue for now I would love to hear your feedback and especially if everything works as expected for your use case! I hope to release the update next week on the App Store.
